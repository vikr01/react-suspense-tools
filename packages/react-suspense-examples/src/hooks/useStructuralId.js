import {useFiber, traverseFiber} from 'its-fine';

export default function useStructuralId(selector) {
    const fiber = useFiber();
    let structuralId = 'foo';
    const stopNode = traverseFiber(
        fiber,
        true,
        function(node, ...args) {
            return selector(node, ...args);
        },
    );
    
    return [structuralId, stopNode];
}