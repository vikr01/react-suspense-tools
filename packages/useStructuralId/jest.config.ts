import type { Config } from "@jest/types";
import baseConfig from "react-suspense-tools-configs/jest.config.base";

const config: Config.InitialOptions = {
  ...baseConfig,
};

export default config;
