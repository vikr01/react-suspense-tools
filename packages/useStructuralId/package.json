{
  "name": "use-structural-id",
  "version": "0.0.0",
  "main": "build/useStructuralId.js",
  "types": "build/useStructuralId.d.ts",
  "author": {
    "name": "Vik R",
    "email": "vik120@icloud.com",
    "url": "http://linkedin.com/in/vikr01/"
  },
  "scripts": {
    "build": "tsup",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "dependencies": {
    "its-fine": "^2.0.0",
    "type-fest": "^4.40.0"
  },
  "devDependencies": {
    "react": "^19.0.0"
  },
  "peerDependencies": {
    "react": "^19.0.0"
  }
}
