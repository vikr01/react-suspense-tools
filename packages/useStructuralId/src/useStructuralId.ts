import {useFiber, traverseFiber, Fiber} from 'its-fine';
import {useRef} from 'react';

type Selector = Parameters<typeof traverseFiber>[2];
export type StructuralId = string;
type StopNode = Fiber | null;

export default function useStructuralId(selector: Selector, dependencies: ReadonlyArray<unknown>): [StructuralId, StopNode] {
    const stopNodeRef = useRef<StopNode>(null);
    const structuralIdRef = useRef<StructuralId>(null);
    const prevDependenciesRef = useRef<ReadonlyArray<unknown>>(null);
    const fiber = useFiber();

    const prevDependencies = prevDependenciesRef.current;
    prevDependenciesRef.current = dependencies;

    if (structuralIdRef.current == null || prevDependencies == null || !arrMatch(prevDependencies, dependencies)) {
        structuralIdRef.current = 'foo';
        stopNodeRef.current = traverseFiber(
            fiber,
            true,
            selector,
            // function(node, ...args) {
            //     return selector(node, ...args);
            // },
        ) ?? null;
    }
    
    return [structuralIdRef.current, stopNodeRef.current];
}

function arrMatch(arr1: ReadonlyArray<unknown>, arr2: ReadonlyArray<unknown>) {
    if (arr1.length !== arr2.length) {
        return false;
    }

    for (let i = 0; i < arr1.length; i++) {
        if (arr1[i] !== arr2[i]) {
            return false;
        }
    }

    return true;
}
