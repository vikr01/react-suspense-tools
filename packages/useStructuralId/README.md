# use-structural-id

Hook to get the structural ID for a React fiber node.

```jsx
<Foo>
    <Bar key={0}>
        <BazFoo />
        <FooFoo /> {/* FooFoo calls useStructuralID */}
    </Bar>
</Foo>
```

The result of calling `useStructuralID` from `FooFoo` will be a string representing the connection: `(FooFoo at index 1)->(Bar with key 0)->(Foo at index 0)`.


## Usage

```javascript
import useStructuralId from 'use-structural-id';

function useMyHook() {
    const structuralId = useStructuralId(
        (fiberNode)=>fiberNode.elementType === Foo, // selector for where to stop the traversal upward, in this case stopping when we reach an element of component type `Foo`
        [], // dependencies
    );
}
```
