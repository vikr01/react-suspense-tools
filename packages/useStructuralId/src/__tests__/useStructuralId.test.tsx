// This is to mock the reactFiberKey that React makes so we can know what it is. We need to do this before anything react is imported.
jest.spyOn(global.Math, 'random').mockReturnValue(0.5);

import * as React from 'react';
import { renderHook } from "@testing-library/react";
import useStructuralId, { type Fiber } from "../useStructuralId";

type Props = {
    children: React.ReactNode,
};

type TestComponent = React.FunctionComponent<Props>;

type ComponentMap = {
    [key: string]: TestComponent,
};

type ReactFiberKey = '_reactFiber$... - random key here, this is just for typescript to properly catch the type';

type ReactFiberHTMLElement = HTMLElement & {[key in ReactFiberKey]: Fiber};

// See here for how react generates this: https://github.com/facebook/react/blob/bc6184dd993e6ea0efdee7553293676db774c3ca/packages/react-dom-bindings/src/client/ReactDOMComponentTree.js#L40
// @ts-expect-error We're forcing the type to _reactFiber since the rest of the string is generated at runtime
const reactFiberKey: ReactFiberKey = `__reactFiber$${Math.random().toString(36).slice(2)}`;

function assertNonNull<T>(x: T): asserts x is NonNullable<T> {
    expect(x).not.toBeNull();
}

describe('useStructuralId', () => {
    let TreeComponent: TestComponent;
    let components: ComponentMap;

    beforeAll(()=>{
        [TreeComponent, components] = generateTree(5);
    });

    afterAll(()=>{
        jest.spyOn(global.Math, 'random').mockRestore();
    });

    it('should return a structural id', () => {
        const {result} = renderHook(()=>useStructuralId(null, []), {wrapper: TreeComponent});

        const [structuralId, stopNode] = result.current;

        expect(structuralId).toMatchInlineSnapshot(`"0:TestComponent:n[0],Component5:n[0],Component4:n[0],Component3:n[0],Component2:n[0],Component1:n[0],Unknown:n[0]"`);

        expect(stopNode).toBeNull();
    });

    it('should find the right fiber node', () => {
        const {result} = renderHook(
            ()=>useStructuralId((node: Fiber)=>node.elementType === components.Component3, []),
            {wrapper: TreeComponent}
        );

        const [structuralId, stopNode] = result.current;

        expect(structuralId).toMatchInlineSnapshot(`"0:TestComponent:n[0],Component5:n[0],Component4:n[0],Component3:n[0]"`);

        expect(stopNode?.elementType).toBe(components.Component3);
    });

    it('should work with html tags', ()=>{
        const OldComponent3 = components.Component3;

        const htmlNodeRef: React.RefObject<ReactFiberHTMLElement | null> = {
            current: null,
        };

        const DivComponent = jest.fn(({children}: Props)=>{
            return (
                <span data-testid="FakeComponent3" ref={htmlNodeRef}>
                    <OldComponent3>
                        {children}
                    </OldComponent3>
                </span>
            );
        });

        components.Component3 = DivComponent;

        const {result} = renderHook(
            ()=>useStructuralId((node: Fiber)=>node.elementType === 'span', []),
            {wrapper: TreeComponent}
        );

        expect(DivComponent).toHaveBeenCalledTimes(1);

        const [structuralId, stopNode] = result.current;

        expect(structuralId.split(',')).toHaveLength(5);

        assertNonNull(htmlNodeRef.current);

        expect(stopNode).toBe(htmlNodeRef.current[reactFiberKey]);

        components.Component3 = OldComponent3;
    });

});

function generateTree(count: number): [TestComponent, ComponentMap] {
    const firstComponentName = `Component${count}`;

    let PrevComponent: TestComponent = ({
        [firstComponentName]({children}: Props) {
            return children;
        }
    })[firstComponentName];

    const components: ComponentMap = {
        [firstComponentName]: PrevComponent,
    };
  
    for (let i = count-1; i >= 1; i--) {
        const name = `Component${i}`;
        
        const Comp: TestComponent = ({
            [name]({children, ...props}: Props) {
                const InnerComponent = components[`Component${i+1}`];

                return (
                    <InnerComponent {...props}>
                        {children}
                    </InnerComponent>
                );
            }
        })[name];

        components[name] = Comp;
        PrevComponent = Comp;
    }

  
    return [PrevComponent, components];
}
