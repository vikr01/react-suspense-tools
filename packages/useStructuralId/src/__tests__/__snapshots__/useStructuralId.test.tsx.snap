// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStructuralId should find the right fiber node 1`] = `"0:TestComponent:n:0,Component5:n:0,Component4:n:0,Component3:n:0"`;

exports[`useStructuralId should return a structural id 1`] = `"0:TestComponent:n:0,Component5:n:0,Component4:n:0,Component3:n:0,Component2:n:0,Component1:n:0,Unknown:n:0"`;
